File: Day-3/Labs/python-kafka-producer-lab/json_order_producer.py
"""
Lab 11- Structured JSON order events (Python).

Run:
    python json_order_producer.py
"""

import json
import sys

from kafka import KafkaProducer
from kafka.errors import KafkaError

from config import DEFAULT_BOOTSTRAP, DEFAULT_TOPIC


def main() -> None:
    bootstrap = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_BOOTSTRAP
    topic = sys.argv[2] if len(sys.argv) > 2 else DEFAULT_TOPIC
    count = int(sys.argv[3]) if len(sys.argv) > 3 else 5

    producer = KafkaProducer(
        bootstrap_servers=bootstrap,
        key_serializer=lambda k: k.encode("utf-8") if k else None,
        value_serializer=lambda v: json.dumps(v).encode("utf-8"),
        acks="all",
    )

    customers = ["Vikash", "John", "Priya", "Amit"]

    try:
        for i in range(1, count + 1):
            order_id = 1000 + i
            event = {
                "orderId": order_id,
                "customer": customers[i % len(customers)],
                "amount": 1000 + i * 500,
                "payment": "UPI",
            }
            key = str(order_id)
            future = producer.send(topic, key=key, value=event)
            try:
                metadata = future.get(timeout=10)
                print(
                    f"Sent key={key} {event} -> "
                    f"partition={metadata.partition} offset={metadata.offset}"
                )
            except KafkaError as exc:
                print(f"Failed: {exc}", file=sys.stderr)
                sys.exit(1)

        producer.flush()
    finally:
        producer.close()

    print("JSON order events sent.")


if __name__ == "__main__":
    main()

File: Day-3/Labs/java-kafka-producer-lab/src/main/java/com/kafka/producer/lab/JsonOrderProducer.java
package com.kafka.producer.lab;

import java.util.Properties;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;

/**
 * Lab 11- Structured JSON order events.
 *
 * <p>Run: mvn -q exec:java -Dexec.mainClass=com.kafka.producer.lab.JsonOrderProducer
 */
public final class JsonOrderProducer {

    public static void main(String[] args) throws Exception {
        String bootstrap = ProducerConfigFactory.arg(args, 0, ProducerConfigFactory.DEFAULT_BOOTSTRAP);
        String topic = ProducerConfigFactory.arg(args, 1, ProducerConfigFactory.DEFAULT_TOPIC);
        int count = ProducerConfigFactory.parseInt(ProducerConfigFactory.arg(args, 2, "5"), 5);

        Properties props = ProducerConfigFactory.baseProps(bootstrap);
        props.put(ProducerConfig.ACKS_CONFIG, "all");
        props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);

        String[] customers = {"Vikash", "John", "Priya", "Amit"};

        try (KafkaProducer<String, String> producer = new KafkaProducer<>(props)) {
            for (int i = 1; i <= count; i++) {
                String customer = customers[i % customers.length];
                int amount = 1000 + i * 500;
                String json = String.format(
                        "{\"orderId\":%d,\"customer\":\"%s\",\"amount\":%d,\"payment\":\"UPI\"}",
                        1000 + i, customer, amount);

                String key = String.valueOf(1000 + i);
                producer.send(new ProducerRecord<>(topic, key, json), (metadata, error) -> {
                    if (error != null) {
                        System.err.println("Failed: " + error.getMessage());
                    } else {
                        System.out.printf(
                                "Sent JSON -> partition=%d offset=%d body=%s%n",
                                metadata.partition(), metadata.offset(), json);
                    }
                }).get();
            }
        }

        System.out.println("JSON messages sent.");
    }
}
